bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/handler/LinkyHandler.java

```diff
@@ -0,0 +1,387 @@
+        LinkyConsumptionData result = cachedYearlyData.getValue();
+        if (result != null && result.success()) {
+            int elementQuantity = result.getData().size();
+            thisYear = elementQuantity > 0 ? result.getData().get(elementQuantity - 1).valeur : -1;
+            lastYear = elementQuantity > 1 ? result.getData().get(elementQuantity - 2).valeur : -1;
+        } else {
+            cachedYearlyData.invalidateValue();
+        }
+        updateKwhChannel(LAST_YEAR, lastYear);
+        updateKwhChannel(THIS_YEAR, thisYear);
+    }
+
+    private void updateKwhChannel(String channelId, double consumption) {
+        logger.debug("Update channel {} with {}", channelId, consumption);
+        updateState(channelId,
+                consumption != -1 ? new QuantityType<>(consumption, SmartHomeUnits.KILOWATT_HOUR) : UnDefType.UNDEF);
+    }
+
+    /**
+     * Produce a report of all daily values between two dates
+     *
+     * @param startDay the start day of the report
+     * @param endDay the end day of the report
+     * @param separator the separator to be used betwwen the date and the value
+     *
+     * @return the report as a string
+     */
+    public String reportValues(LocalDate startDay, LocalDate endDay, @Nullable String separator) {
+        String dump = "";
+        if (startDay.getYear() == endDay.getYear() && startDay.getMonthValue() == endDay.getMonthValue()) {
+            // All values in the same month
+            LinkyConsumptionData result = getConsumptionData(DAILY, startDay, endDay, true);
+            if (result != null && result.success()) {
+                LocalDate currentDay = startDay;
+                int jump = result.getDecalage();
+                while (jump < result.getData().size() && !currentDay.isAfter(endDay)) {
+                    double consumption = result.getData().get(jump).valeur;
+                    dump += currentDay.format(DateTimeFormatter.ISO_LOCAL_DATE) + separator;
+                    if (consumption >= 0) {
+                        dump += String.valueOf(consumption);
+                    }
+                    dump += "\n";
+                    jump++;
+                    currentDay = currentDay.plusDays(1);
+                }
+            } else {
+                LocalDate currentDay = startDay;
+                while (!currentDay.isAfter(endDay)) {
+                    dump += currentDay.format(DateTimeFormatter.ISO_LOCAL_DATE) + separator + "\n";
+                    currentDay = currentDay.plusDays(1);
+                }
+            }
+        } else {
+            // Concatenate the report produced for each month between the two dates
+            LocalDate first = startDay;
+            do {
+                LocalDate last = first.withDayOfMonth(first.lengthOfMonth());
+                if (last.isAfter(endDay)) {
+                    last = endDay;
+                }
+                dump += reportValues(first, last, separator);
```

Comment:
Surely there is a way to implement this without recursion right?

Reply:
I would keep recursion also.